---
title: "DNA methylation analysis using bisulfite sequencing data"
output: html_notebook
---

# DNA methylation
Cytosine methylation (5-methylcytosine, 5mC) is one of the main covalent base modifications in eukaryotic genomes, generally observed on CpG dinucleotides. 
Methylation can also rarely occur in a non-CpG context, but this was mainly observed in human embryonic stem and neuronal cells (Lister, Pelizzola, Dowen, et al. 2009; 
Lister, Mukamel, Nery, et al. 2013). DNA methylation is a part of the epigenetic regulation mechanism of gene expression. It is cell-type-specific DNA modification. It is reversible but mostly 
remains stable through cell division. There are roughly 28 million CpGs in the human genome, 60–80% are generally methylated. Less than 10% of CpGs occur in CG-dense regions that are termed 
CpG islands in the human genome (Smith and Meissner 2013). It has been demonstrated that DNA methylation is also not uniformly distributed over the genome, but rather is associated with CpG 
density. In vertebrate genomes, cytosine bases are usually unmethylated in CpG-rich regions such as CpG islands and tend to be methylated in CpG-deficient regions. Vertebrate genomes are 
largely CpG deficient except at CpG islands. Conversely, invertebrates such as Drosophila melanogaster and Caenorhabditis elegans do not exhibit cytosine methylation and consequently do not 
have CpG rich and poor regions but rather a steady CpG frequency over their genomes (Deaton and Bird 2011).

DNA methylation is established by DNA methyltransferases DNMT3A and DNMT3B in combination with DNMT3L and maintained through cell division by the methyltransferase DNMT1 and associated proteins. 
DNMT3a and DNMT3b are in charge of the de novo methylation during early development. Loss of 5mC can be achieved passively by dilution during replication or exclusion of DNMT1 from the nucleus. 
Recent discoveries of the ten-eleven translocation (TET) family of proteins and their ability to convert 5-methylcytosine (5mC) into 5-hydroxymethylcytosine (5hmC) in vertebrates provide a path 
for catalyzed active DNA demethylation (Tahiliani, Koh, Shen, et al. 2009). Iterative oxidations of 5hmC catalyzed by TET result in 5-formylcytosine (5fC) and 5-carboxylcytosine (5caC). 5caC 
mark is excised from DNA by G/T mismatch-specific thymine-DNA glycosylase (TDG), which as a result reverts cytosine residue to its unmodified state (He, Li, Li, et al. 2011). Apart from these, 
mainly bacteria, but possibly higher eukaryotes, contain base modifications on bases other than cytosine, such as methylated adenine or guanine (Clark, Spittle, Turner, et al. 2011).

# Bisulfite sequencing
This method, and the related ones, allow measurement of DNA methylation at the single nucleotide resolution. The bisulfite conversion turns unmethylated Cs to Ts and methylated Cs remain intact. 
Then, the only thing to do is to align the reads with those C->T conversions and count C->T mutations to calculate fraction of methylated bases. In the end, we can get quantitative genome-wide 
measurements for DNA methylation.

See https://compgenomr.github.io/book/what-is-dna-methylation.html

## Install methylKit if not already
```{R}
BiocManager::install("methylKit")
```

## Import libraries
```{R}
library(methylKit)
```

```{R}
file.list=list(
system.file("extdata", "test1.myCpG.txt", package = "methylKit"),
system.file("extdata", "test2.myCpG.txt", package = "methylKit"),
system.file("extdata", "control1.myCpG.txt", package = "methylKit"),
system.file("extdata", "control2.myCpG.txt", package = "methylKit")
)


myobj=methRead(file.list,
           sample.id=list("test1","test2","ctrl1","ctrl2"),
           assembly="hg18",
           treatment=c(1,1,0,0),
           context="CpG"
           )

```

```{R}
length(myobj)
head(myobj[[1]])
```
```{R}
getMethylationStats(myobj[[2]],plot=TRUE,both.strands=FALSE)
```

```{R}
getCoverageStats(myobj[[2]],plot=TRUE,both.strands=FALSE)
```

## Filter samples based on coverage
```{R}
filtered.myobj=filterByCoverage(myobj,lo.count=10,lo.perc=NULL,
                                      hi.count=NULL,hi.perc=99.9)
```

## Merging samples into a single table
```{R}
meth=methylKit::unite(myobj, destrand=FALSE)
head(meth)

# creates a methylBase object, 
# where only CpGs covered with at least 1 sample per group will be returned

# there were two groups defined by the treatment vector, 
# given during the creation of myobj: treatment=c(1,1,0,0)
# meth.min=unite(myobj,min.per.group=1L)
```

## Filtering CpGs
We might need to filter the CpGs further before exploratory analysis or even before the downstream analysis such as differential methylation. 
For exploratory analysis, it is of general interest to see how samples relate to each other and we might want to remove CpGs that are not variable 
before doing that. Or we might remove Cs that are potentially C->T mutations. First, we show how to filter based on variation. Below, we extract 
percent methylation values from CpGs as a matrix. Calculate the standard deviation for each CpG and filter based on standard deviation. 

```{R}
pm=percMethylation(meth) # get percent methylation matrix
mds=matrixStats::rowSds(pm) # calculate standard deviation of CpGs
head(meth[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")
```

Now, let’s assume we know the locations of C->T mutations. These locations should be removed from the analysis as they do not represent 
bisulfite-treatment-associated conversions. Mutation locations are stored in a GRanges object, and we can use that to remove CpGs overlapping 
with mutations. In order to do the overlap operation, we will convert the methylKit object to a GRanges object and do the filtering with the 
%over% function within [ ]. The returned object will still be a methylKit object.

```{R}
library(GenomicRanges)
# example SNP
mut=GRanges(seqnames=c("chr21","chr21"),
            ranges=IRanges(start=c(9853296, 9853326),
                           end=c( 9853296,9853326)))

# select CpGs that do not overlap with mutations
sub.meth=meth[! as(meth,"GRanges") %over% mut,]
paste("nrow meth", nrow(meth))
paste("nrow sub.meth", nrow(sub.meth))
```

## Dendrogram
```{R}
clusterSamples(meth, dist="correlation", method="ward", plot=TRUE)
```
## PCA
```{R}
PCASamples(meth, screeplot=TRUE)
pc=PCASamples(meth,obj.return = TRUE, adj.lim=c(1,1))
```

